File: PDOO-Napakalaki/Napakalaki/NapakalakiRuby/lib/napakalaki.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require "singleton"
require_relative "prize.rb"
require_relative "monster.rb"
require_relative "bad_consequence.rb"
require_relative "treasure.rb"
require_relative "treasure_kind.rb"
require_relative "card_dealer.rb"
require_relative "player.rb"
require_relative "combat_result.rb"
require_relative "cultist.rb"
require_relative "cultist_player.rb"

class Napakalaki
    include Singleton
    
    attr_accessor :currentPlayer, :players, :dealer, :currentMonster
  
   
    
    def initialize
      @currentPlayer = nil
      @players = Array.new
      @dealer = CardDealer.getInstance
      @currentMonster = nil
    end
    
    def getCurrentPlayer 
      return @currentPlayer
    end
    
    def getCurrentMonster
      return @currentMonster
    end
    
    def initPlayers(names) 
        names.each { |s|        
          @players << Player.new(s)
        }
    end
    
    def nextPlayer
      
       if (@currentPlayer == nil) 
                  
            sig = rand(@players.length)
            @currentPlayer = @players.at(sig)
      
      else
            sig = 0
            for i in (0..(@players.length-1))
            
                if(@players.at(i).name == @currentPlayer.name) 
                    sig = i                  
                end
            end
            sig = (sig + 1) % @players.length
            @currentPlayer = @players.at(sig)
       end  
                          
        return @currentPlayer
    end
    
    def nextTurnAllowed
        
       if(@currentPlayer == nil) 
            return true
              
        else 
            return @currentPlayer.validState
       end
        
    end
    
    def setEnemies        
      
        for i in (0..(@players.length-1)) 
            
            newEnemy = rand(@players.length)
            while(@players.at(i).enemy == nil) do
                newEnemy = rand(@players.length)
                if((newEnemy != i) && (@players.at(i).enemy == nil)) 
                    @players.at(i).enemy = @players.at(newEnemy)
                end
            end
        end
        
    end
    
    def self.getInstance
      return Napakalaki.instance
    end  
    
    def developCombat
      
       combatResult = @currentPlayer.combat(@currentMonster)
       @dealer.giveMonsterBack(@currentMonster)
       
      if(combatResult == CombatResult::LOSEANDCONVERT)
        c = @dealer.nextCultist
        p = CultistPlayer.new(@currentPlayer, c)
        @currentPlayer = p        
        for i in (0..(@players.length-1))
            
                if(@players.at(i).name == @currentPlayer.name) 
                    @players[i] = p              
                end
        end        
      end
      
       return combatResult
    end
    
    def discardVisibleTreasures(treasures)
        
      treasures.each { |t|
        
        @currentPlayer.discardVisibleTreasure(t)
        @dealer.giveTreasureBack(t)
      }     
        
    end
    
    def discardHiddenTreasures(treasures) 
        
      treasures.each { |t|
        
        @currentPlayer.discardHiddenTreasure(t)
        @dealer.giveTreasureBack(t)
      } 
      
    end
    
    def makeTreasuresVisible(treasures) 
        
      treasures.each { |t|
        
        @currentPlayer.makeTreasureVisible(t)
      }
            
    end
            
    def initGame(players)
        
        initPlayers(players)
        setEnemies
        @dealer.initCards
        nextTurn
    end
                       
    def nextTurn
     
        stateOk = nextTurnAllowed       
        if(stateOk) 
            @currentMonster = @dealer.nextMonster
            @currentPlayer = nextPlayer
            dead = @currentPlayer.isDead
            if(dead)
                @currentPlayer.initTreasures
            end
        end
        
        return stateOk
        
    end
            
    def endOfGame(result) 
    
      if(result == CombatResult::WINGAME) 
            return true
              
        else 
            return false
       end
    
    end
    private :initPlayers, :nextPlayer, :nextTurnAllowed, :setEnemies
end

File: PDOO-Napakalaki/Napakalaki/NapakalakiJava/src/NapakalakiGame/Napakalaki.java

package NapakalakiGame;

import java.util.ArrayList;
import java.util.Random;

public class Napakalaki {
    
    private static final Napakalaki instance=new Napakalaki();
    private CombatResult result;
    private Player currentPlayer;
    
    private ArrayList<Player>players=new ArrayList();
    private Monster currentMonster;
    private CardDealer dealer;
    
    Napakalaki(){}

    private void initPlayers(ArrayList<String>names){
        
        for(String tmp : names){
            Player aux=new Player(tmp);
            players.add(aux);
        }
    }
    
     private Player nextPlayer(){
       
       int next;
       Player nextplayer;
       
       int num_total = players.size();
       
       if(currentPlayer == null){
       
           Random random = new Random();
           
           next = random.nextInt(num_total);
       }
       
       else{
       
           int next_current = players.indexOf(currentPlayer);
           
           if(next_current == num_total-1){
           
               next = 0;
           }
           
           else{
           
               next = next_current+1;
           }
       }
       
       nextplayer = players.get(next);
       currentPlayer = nextplayer;
       
       return currentPlayer;
   }
    
    private boolean nextTurnAllowed(){
        
        return (this.currentPlayer==null) || (this.currentPlayer.validState()==true);
    }
    
    private void setEnemies(){
      for(Player tmp : this.players){
          
        int indice=(int)(Math.random()*(players.size()-1));
      
        while(indice==players.indexOf(this))
            indice=(int)(Math.random()*(players.size()-1));
      
        tmp.setEnemy(players.get(indice));
       }
    }
    
    public static Napakalaki getInstance(){return instance;}
    
    public CombatResult developCombat(){
        
        CombatResult combat;
        Monster m=this.currentMonster;
        dealer=CardDealer.getInstance();
        
        combat=this.currentPlayer.combat(m);
        dealer.giveMonsterBack(m);
        
        if(combat==CombatResult.LOSEANDCONVERT){
            Cultist tmp;
            tmp=dealer.nextCultist();
            
            CultistPlayer nplayer = new CultistPlayer(currentPlayer,tmp);
            
            int index = players.indexOf(currentPlayer);
            this.players.set(index, currentPlayer);
            this.currentPlayer=nplayer;
        }
        
        return combat;
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure>treasures){
        
        for(Treasure tmp : treasures){
            this.currentPlayer.discardVisibleTreasure(tmp);
            dealer.giveTreasureBack(tmp);
        }
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure>treasures){
    
        for(Treasure tmp : treasures){
            this.currentPlayer.discardHiddenTreasure(tmp);
            dealer.giveTreasureBack(tmp);
        }
    }
    
    public void makeTreasuresVisible(ArrayList<Treasure>treasures){
        
        for(Treasure tmp : treasures){
            this.currentPlayer.makeTreasureVsible(tmp);
        }
    
    }
    
    public void initGame(ArrayList<String>players){

        this.initPlayers(players);
        this.setEnemies();

        dealer = CardDealer.getInstance();
        dealer.initCards();        

        this.nextTurn();
    }
    
    public Player getCurrentPlayer(){return currentPlayer;}
    
    public Monster getCurrentMonster(){return currentMonster;}
   
    public boolean nextTurn(){
        
        boolean stateOK=this.nextTurnAllowed();
        dealer=CardDealer.getInstance();
        
        if(stateOK){
            
            this.currentMonster=dealer.nextMonster();
            this.currentPlayer=this.nextPlayer();
            
            boolean death=this.currentPlayer.isDead();
            
            if(death){
                this.currentPlayer.initTreasures();
            }
        }
    return stateOK;
    }
    
    public boolean endOfGame(CombatResult result){
        
        return this.result==CombatResult.WINGAME;
    
        }

    @Override
    public String toString() {
        return "Napakalaki{" + "result=" + result + ", currentPlayer=" + currentPlayer + ", players=" + players + ", currentMonster=" + currentMonster + ", dealer=" + dealer + '}';
    }
    
}